base/src/main/java/com/mindera/skeletoid/widgets/tooltip/PopupTooltip.kt

```diff
@@ -22,270 +23,186 @@ import androidx.annotation.DrawableRes
-    fun respectScreenBounds(clipToChildren: Boolean) = apply { this.respectScreenBounds = clipToChildren }
-
-    /**
-     * The tooltip width to be displayed in dips (Density Independent Pixels).
-     * 250dp seems to be a good roundabout value.
-     *
-     * @param width - the tooltip width in dips.
-     */
-    fun layoutWidth(width: Int) = apply { this.popupWidth = width }
-
-    /**
-     * Defines a custom layout for this tooltip.
-     */
-    fun layout(@LayoutRes layout: Int) = apply { this.layout = layout }
-
-    /**
-     * Defines a custom elevation for all the tooltip components, in dips (Density Independent Pixels).
-     *
-     * @param elevation - the amount of elevation in dips.
-     */
-    fun elevation(elevation: Int) = apply { this.elevation = elevation }
+    fun clearTooltip() {
+        tooltip?.dismissIfShowing()
+        tooltip = null
+    }
 
     /**
      * Creates a touch delegate that increases the bounds of the view's touch area by a given amount, in
      * dips (Density Independent Pixels).
      *
      * @param increaseAmount - the amount to increase in dips.
      */
-    fun touchDelegateTouchAreaIncreaseAmount(increaseAmount: Int) = apply { setupTouchDelegate(anchorView, increaseAmount, anchorView.context) }
-
-    /**
-     * Defines the arrow height in dips (Density Independent Pixels).
-     *
-     * @param height - the arrow height in dips.
-     */
-    fun arrowHeight(height: Int) = apply { this.arrowHeight = height }
-
-    /**
-     * Defines the arrow width in dips (Density Independent Pixels).
-     *
-     * @param width - the arrow width in dips.
-     */
-    fun arrowWidth(width: Int) = apply { this.arrowHeight = width }
+    fun touchDelegateTouchAreaIncreaseAmount(increaseAmount: Int) =
+        apply { setupTouchDelegate(anchorView, increaseAmount, anchorView.context) }
 
     /**
      * Displays the tooltip
      */
-    fun show() = setupTooltip(anchorView, message)
+    fun show() {
+        showTooltip(anchorView, message)
+    }
 
     /**
      * Internal helper class to show the tooltip.
      */
-    @SuppressLint("InflateParams")
-    private fun setupTooltip(view: View, text: CharSequence?) {
+    private fun showTooltip(anchorView: View, message: CharSequence) {
         handler.removeCallbacksAndMessages(null)
 
         if (tooltip != null) {
             clearTooltip()
             return
         }
 
-        if (TextUtils.isEmpty(text)) {
+        if (TextUtils.isEmpty(message)) {
             return
         }
 
+        val (x, y) = setupTooltip(anchorView, message)
+
+        tooltip?.showAtLocation(anchorView, GRAVITY, x.toInt(), y.toInt())
+    }
+
+    /**
+     * Internal helper class to create the tooltip and calculate the display coordinates.
+     */
+    @SuppressLint("InflateParams")
+    private fun setupTooltip(view: View, text: CharSequence?): Coordinates {
```

Comment:
This function seems to be doing a lot. It's called `setupTooltip` but returns a `Coordinates` object which sounds like a side effect ?!

Can we try to break it into smaller chunks?

Reply:
Fair point..I'll try to split the setup and the calculations into two different functions, that makes more sense yeah